include/aspect/material_model/averaging.h

```diff
@@ -0,0 +1,148 @@
+         * inputs in @p in.
+         */
+        virtual
+        void
+        evaluate (const typename Interface<dim>::MaterialModelInputs &in,
+                  typename Interface<dim>::MaterialModelOutputs &out) const;
+        /**
+         * Method to declare parameters related to depth-dependent model
+         */
+        static void
+        declare_parameters (ParameterHandler &prm);
+
+        /**
+         * Method to parse parameters related to depth-dependent model
+         */
+        virtual void
+        parse_parameters (ParameterHandler &prm);
+
+        /**
+         * Method to determine viscosity dependencies of material model
+         * These are derived from the base model.
+         */
+        virtual bool
+        viscosity_depends_on (const NonlinearDependence::Dependence dependence) const;
+
+        /**
+         * Method to determine density dependencies of material model
+         * These are derived from the base model.
+         */
+        virtual bool
+        density_depends_on (const NonlinearDependence::Dependence dependence) const;
+
+        /**
+         * Method to determine compressibility dependencies of material model
+         * These are derived from the base model.
+         */
+        virtual bool
+        compressibility_depends_on (const NonlinearDependence::Dependence dependence) const;
+
+        /**
+         * Method to determine specific heat capacity dependencies of material model
+         * These are derived from the base model.
+         */
+        virtual bool
+        specific_heat_depends_on (const NonlinearDependence::Dependence dependence) const;
+
+        /**
+         * Method to determine thermal conductivity dependencies of material model
+         * These are derived from the base model.
+         */
+        virtual bool
+        thermal_conductivity_depends_on (const NonlinearDependence::Dependence dependence) const;
+
+        /**
+         * Method that indicates whether material is compressible. Depth dependent model is compressible
+         * if and only if base model is compressible.
+         */
+        virtual bool is_compressible () const;
+
+        /**
+         * Method to calculate reference viscosity for the depth-dependent model. The reference
+         * viscosity is determined by evaluating the depth-dependent part of the viscosity at
+         * the mean depth of the model.
+         */
+        virtual double reference_viscosity () const;
+
+        /**
+         * Method to calculate reference density for the depth-dependent model. Because the depth-
+         * dependent model deos not modify density, the reference density is equivalent to the
+         * base model's reference density.
+         */
+        virtual double reference_density () const;
+        /**
+                 * Bellshape limit
+                 */
+        double bell_shape_limit;
+        std::string averaging_operation;
```

Comment:
also, I think you should introduce an enum that stores the different averaging operations. The `averaging_operation` variable would then be of this enum type, and you set it to its correct value based on the string variable you get in `parse_parameters()`.

This has the advantage that you don't need to do string comparisons in `evaluate()` -- only integer comparisons with the enum necessary there, which you could even do with a `switch-case` statement.